dali/python/nvidia/dali/shared_batch.py

```diff
@@ -0,0 +1,221 @@
+# Copyright (c) 2020, NVIDIA CORPORATION. All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from nvidia.dali import shared_mem, tensors
+from nvidia.dali import types
+import pickle
+
+
+np = None
+
+
+def import_numpy():
+    global np
+    if np is None:
+        try:
+            import numpy as np
+        except ImportError:
+            raise RuntimeError('Could not import numpy. Please make sure you have numpy '
+                               'installed before you use parallel mode.')
+
+
+class SampleMeta:
+    """Metadata describing serialized sample in a memory buffer.
+
+    It is passed through memory, stored after sample it describes."""
+
+    def __init__(self, offset, shape, dtype, nbytes):
+        self.shape = shape
+        self.dtype = dtype
+        self.offset = offset
+        self.nbytes = nbytes
+
+    @classmethod
+    def from_np(cls, offset, np_array):
+        return cls(offset, np_array.shape, np_array.dtype, np_array.nbytes)
+
+
+class SharedBatchMeta:
+    """Container for essential meta data about batch written into shared memory.
+    Contains id of the memory chunk, size of the buffer, and offset of more detailed
+    meta data passed through shared memory (SampleMeta).
+    """
+
+    def __init__(self, mem_chunk_id, capacity, meta_offset, meta_size):
+        self.mem_chunk_id = mem_chunk_id
+        self.capacity = capacity
+        self.meta_offset = meta_offset
+        self.meta_size = meta_size
+
+    @classmethod
+    def from_writer(cls, writer):
+        return cls(writer.mem_batch.mem_chunk_id, writer.mem_batch.capacity, writer.data_size,
+                   writer.meta_data_size)
+
+
+def deserialize_sample(buffer: shared_mem.SharedMem, sample):
+    if isinstance(sample, SampleMeta):
+        offset = sample.offset
+        buffer = buffer.buf[offset:offset + sample.nbytes]
+        return np.ndarray(sample.shape, dtype=sample.dtype, buffer=buffer)
+    if isinstance(sample, (tuple, list,)):
+        return type(sample)(deserialize_sample(buffer, part) for part in sample)
```

Comment:
Why not just:
```suggestion
        return [deserialize_sample(buffer, part) for part in sample]
```
Do we really need to return tuple or list respectively or just any iterable?

Reply:
It's the multiple outputs handling, tbh probably we don't need it. If anything I would probably go with tuple, or leave it as is. I will add a test that list works. I don't think anyone would return other iterable (I hope).